dataavailability/dataavailability.go

```diff
@@ -0,0 +1,117 @@
+
+	ctx context.Context
+}
+
+// New creates a DataAvailability instance
+func New(
+	IsTrustedSequencer bool,
+	backend DABackender,
+	state stateInterface,
+	zkEVMClient syncinterfaces.ZKEVMClientInterface,
+) (*DataAvailability, error) {
+	da := &DataAvailability{
+		isTrustedSequencer: IsTrustedSequencer,
+		backend:            backend,
+		state:              state,
+		zkEVMClient:        zkEVMClient,
+		ctx:                context.Background(),
+	}
+	err := da.backend.Init()
+	return da, err
+}
+
+// PostSequence sends the sequence data to the data availability backend, and returns the dataAvailabilityMessage
+// as expected by the contract
+func (d *DataAvailability) PostSequence(ctx context.Context, sequences []types.Sequence) ([]byte, error) {
+	batchesData := [][]byte{}
+	for _, batch := range sequences {
+		// Do not send to the DA backend data that will be stored to L1
+		if batch.ForcedBatchTimestamp == 0 {
+			batchesData = append(batchesData, batch.BatchL2Data)
+		}
+	}
+	return d.backend.PostSequence(ctx, batchesData)
+}
+
+// GetBatchL2Data tries to return the data from a batch, in the following priorities
+// 1. From local DB
+// 2. From Sequencer
+// 3. From DA backend
+func (d *DataAvailability) GetBatchL2Data(batchNum uint64, expectedTransactionsHash common.Hash) ([]byte, error) {
+	found := true
+	transactionsData, err := d.state.GetBatchL2DataByNumber(d.ctx, batchNum, nil)
+	if err != nil {
+		if err == state.ErrNotFound {
+			found = false
+		} else {
+			return nil, fmt.Errorf("failed to get batch data from state for batch num %d: %w", batchNum, err)
+		}
+	}
+	actualTransactionsHash := crypto.Keccak256Hash(transactionsData)
+	if !found || expectedTransactionsHash != actualTransactionsHash {
+		if found {
+			log.Warnf(unexpectedHashTemplate, batchNum, expectedTransactionsHash, actualTransactionsHash)
+		}
+
+		if !d.isTrustedSequencer {
+			log.Info("trying to get data from trusted sequencer")
+			data, err := d.getDataFromTrustedSequencer(batchNum, expectedTransactionsHash)
+			if err != nil {
+				log.Error(err)
+			} else {
+				return data, nil
+			}
+		}
+
+		log.Info("trying to get data from the data availability backend")
+		data, err := d.backend.GetData(batchNum, expectedTransactionsHash)
+		if err != nil {
+			log.Error(err)
+			if d.isTrustedSequencer {
+				return nil, fmt.Errorf("data not found on the local DB nor on any data committee member")
+			} else {
+				return nil, fmt.Errorf("data not found on the local DB, nor from the trusted sequencer nor on any data committee member")
+			}
+		}
+		return data, nil
+	}
+	return transactionsData, nil
+}
+
+const unexpectedHashTemplate = "missmatch on transaction data for batch num %d. Expected hash %s, actual hash: %s"
+
+func (d *DataAvailability) getDataFromTrustedSequencer(batchNum uint64, expectedTransactionsHash common.Hash) ([]byte, error) {
+	b, err := d.zkEVMClient.BatchByNumber(d.ctx, big.NewInt(int64(batchNum)))
+	if err != nil {
```

Comment:
Is it safe to do down-casting from `uint64` to `int64` here? I'd suggest safer option, wdyt?

```suggestion
	b, err := d.zkEVMClient.BatchByNumber(d.ctx, new(big.Int).SetUint64(batchNum))
```